BITSWAP.md

```diff
@@ -196,6 +197,94 @@ message Message {
 }
 ```
 
+## Bitswap 1.3.0
+
+Bitswap 1.3.0 extends the Bitswap 1.2.0 protocol with the following changes:
+1. Having a list of tokens that may be sent with the message and referenced within the message
+2. Allowing each entry in the wantlist to contain a set of tokens
```

Comment:
From @ianopolous:

> Do you have a concrete use case in mind for allowing multiple auth tokens for a single cid in the same request? Our usage in Peergos is maximally fine grained and doesn't need this. For example, Amazon S3 doesn't need this and they clearly support almost all use cases in the world right now.

A couple things come to mind:
- Requiring two tokens from different systems: e.g. a Peergos token for authentication and some token related to paying for retrieval of that data
- "Suggested" ACLs: The server wants to tell the client which token caused me to give them the block (e.g. for multitenancy auth or payment purposes), but I also want to tell the client some general ACL I'd like them to respect

Maybe it's overkill, but we're making it easier for people to do auth (e.g. they could've just used peerID auth and some offline channel to coordinate authentication and payments) and we should probably try to make systems composable in-band. Thoughts?

Reply:
Separating the auth token from a payment token is an interesting idea. Has anyone done anything like this (is your angle composability with filecoin?)? I would normally imagine you would just pay for access to the content. I.e. pay to get an auth token/key. But I'm just hypothesising here. I guess what you're saying is different because you're not paying for access, you're paying for delivery, which should be independent. 

Note that in our multi-tenant case, the client already knows which token was authed to get a block without needing this.

Reply:
Use cases I have talked to folks about: 
- ACLs without payments
- ACLs with payments
- request limiter, e.g. this token allows you to retrieve 100 times

Reply:
To be clear, one absolutely critical property of any usage of this is that auth tokens MUST be tied to the sender's nodeId and this must be verified before releasing a block. Otherwise you are trivially vulnerable to replay attacks because you broadcast the tokens to the DHT. I think it would be worthwhile to flesh out a few of these use cases to make sure it is possible to do securely. E.g. a naive approach of using this to do an ACL by having tokens that the recipient looks up in a db to check that a token is allowed for a cid is insecure. To make it secure the db would also need to store the nodeId's allowed to use a given token, but then it wouldn't need this extension at all or any auth tokens, just use the nodeid of the requestor in normal bitswap.

Reply:
The more I think about trying to use this to do ACLs (as opposed to capabilities) the more I think it is impossible to do securely because of the broadcast nature of bitswap, except in the trivial case where the ACL only depends on the nodeId. In that case you don't need this PR at all and you should just use a custom allow function as suggested [here](https://github.com/ipfs/go-ipfs/issues/7871) and as we've implemented [here](https://github.com/Peergos/go-bitswap-auth/blob/master/auth/blockstore.go#L42).

Fundamentally, it comes down to ACLs depending on identity and the only identity between ipfs nodes is the node id. So the ACL will only be able to depend on the node id, in which case why not just use the node id which is already authenticated and unfakeable without this PR. 

There is one way around this though, and that is single-use (not N>1) tokens. If a token is enforced to be only used once, then clearly a replay attack can't occur. That sounds like a massive footgun to me and also a scalabilty and centralisation problem as well.

Reply:
Scratch my last statement about single use tokens, because whilst true, there is also a race condition between the originator and anyone else in the DHT they send the token to. So yes that is a footgun.

Reply:
Bitswap doesn't have a mechanism to only send wants to "the correct" peers. From a security perspective Wants are broadcast to the world. (If I want to receive an auth token for a given cid then I just need to pretend to the DHT that I have that cid and then the requests will come in with valid tokens). This is super important, and will trivially break security for most things people will try and do with it. 

It is possible to do securely *with* the broadcast assumption, which is what we've [figured out](https://peergos.org/posts/bats) in Peergos. A simplified analogy would be that if a secret S allows access to cid C. Then if node A is requesting C then the token sent with bitswap could be some form of signature using S of (A, C). Then the recipient can verify that the signature is valid, and the request came from A. Then there is no possibility for replay attacks because the signature is tied to the sender's node id. 

So I highly recommend speccing out some of these proposed usages and checking they are not totally broken security-wise.

Reply:
> Bitswap doesn't have a mechanism to only send wants to "the correct" peers

I don't understand this. Bitswap is a message sending protocol between two peers. 

> From a security perspective Wants are broadcast to the world.

This is only the case in go-bitswap, the decision as to which peers to send messages to is up to the implementation. Is there somewhere in the spec that suggests that wants are globally broadcast?

Security aside implementations may not want to broadcast to everyone simply as a way of cutting down on spamming and bandwidth usage. For example, even in the non-auth case I could see an implementation making the argument to only ask peers it's connected to over a LAN or where they were recommended by the content routing system.

> So I highly recommend speccing out some of these proposed usages and checking they are not totally broken security-wise.

A couple that are more suitable for open networks have been described (e.g. bats and payments), but doing simple bearer token style auth seems very easy as well. It just means that writing some CLI command like `fetch-cid <cid> <bearer-token>` is meaningless and instead you'd need to specify `fetch-cid <cid> <multiaddr> [bearer-token]` or `fetch-cid <cid> [tuple of mulitaddr and bearer token]`.

I'll try and reach out to some of the folks that have requested this functionality in the past for comments since AFAIK none of the people who have commented on this PR are planning on implementing support for something closer to a bearer-token style which wouldn't be appropriate for broadcast.

However, I'm confused as to what this has to do with the spec proposal. Even bats + payments require a spec change that looks like this though, right?

Reply:
> This is only the case in go-bitswap
To my knowledge, go-bitswap is the main implementation of bitswap, which is used by go-ipfs the main implementation of ipfs. Are you suggesting that existing ipfs apis wouldn't use this at all? If only new apis like you suggest will use this then that's much easier to avoid footguns.